File: core/clientLaya/game/src/shine/support/quartz/IndexMaker.ts
namespace Shine
{
	export class IndexMaker
	{
	    private _min:number;
	
		private _max:number;
	
		private _canRound:boolean;
	
		private _index:number;
	
		constructor(min:number=0,max:number=2147483646,canRound:boolean=false)
		{
			this._min=min;
			this._max=max;
			this._canRound=canRound;
	
			this._index=this._min;
		}
	
		/** 重置 */
		public reset():void
		{
			this._index=this._min;
		}
	
		/** 取一个序号 */
		public get():number
		{
			var re:number=++this._index;
	
			if(re >= this._max)
			{
				if(this._canRound)
				{
					this._index=this._min;
					re=++this._index;
				}
				else
				{
					Ctrl.throwError("序号构造溢出");
				}
			}
	
			return re;
		}
	}
}

File: core/server/project/shine/src/main/java/com/home/shine/support/IndexMaker.java
package com.home.shine.support;

import com.home.shine.ctrl.Ctrl;

/** 序号构造器(上下限均不会达到)(min<value<max) */
public class IndexMaker
{
	private int _min;
	private int _max;
	
	private boolean _canRound;
	
	private int _index;
	
	public IndexMaker()
	{
		this(1,Integer.MAX_VALUE,false);
	}
	
	public IndexMaker(int min,int max,boolean canRound)
	{
		_min=min;
		_max=max;
		_canRound=canRound;
		
		_index=_min;
	}
	
	public void setMin(int min)
	{
		_min=min;
		
		_index=_min;
	}
	
	/** 重置 */
	public void reset()
	{
		_index=_min;
	}
	
	/** 取一个序号 */
	public int get()
	{
		int re=++_index;

		if(re >= _max)
		{
			if(_canRound)
			{
				_index=_min;
				re=++_index;
			}
			else
			{
				Ctrl.throwError("序号构造溢出");
			}
		}

		return re;
	}
}
